File: src/discord.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"
)

type DiscordPayload struct {
	Embeds []DiscordEmbed `json:"embeds"`
}

type DiscordEmbed struct {
	Title  string         `json:"title"`
	Fields []DiscordField `json:"fields"`
	Color  int            `json:"color"`
}

type DiscordField struct {
	Name   string `json:"name"`
	Value  string `json:"value"`
	Inline bool   `json:"inline"`
}

type DiscordResponse struct {
	Id string `json:"id"`
}

func formatDuration(seconds float64) string {
	duration := time.Duration(int(seconds * 1e9))
	return fmt.Sprintf("%s", duration)
}

const progressBarLength = 20

func formatProgressBar(ratio float64) string {
	var buffer bytes.Buffer
	for i := 0; i < progressBarLength; i++ {
		p := float64(i) / progressBarLength
		if p >= ratio {
			buffer.WriteRune('░')
		} else {
			buffer.WriteRune('█')
		}
	}
	buffer.WriteString(
		fmt.Sprintf(" %.2f%%", ratio*100),
	)
	return buffer.String()
}

func makePayloadBuffer(task *Task) *bytes.Buffer {
	fields := []DiscordField{
		{Name: "Path", Value: task.Path, Inline: false},
		{Name: "Output path", Value: task.OutputPath, Inline: false},
	}
	if task.Status == PROCESSING && task.Probe != nil {
		seconds, _ := strconv.ParseFloat(task.Probe.Format.Duration, 64)
		fields = append(fields, DiscordField{
			Name: "Duration", Value: formatDuration(seconds), Inline: true,
		})
		if task.Progress != nil {
			fields = append(fields, DiscordField{
				Name: "Timestamp", Value: formatDuration(task.Progress.Timestamp), Inline: true,
			}, DiscordField{
				Name: "Speed", Value: task.Progress.Speed, Inline: true,
			}, DiscordField{
				Name: "Progress", Value: formatProgressBar(
					task.Progress.Timestamp / seconds,
				), Inline: false,
			})
		}
	}
	var color int
	switch task.Status {
	case QUEUED:
		color = 0xa855f7
	case PROCESSING:
		color = 0xf97316
	case DONE:
		color = 0x22c55e
	case FAILED:
		color = 0xef4444
	default:
		color = 0x737373
	}
	payload := DiscordPayload{
		Embeds: []DiscordEmbed{
			{
				Title:  "Transcoding file",
				Fields: fields,
				Color:  color,
			},
		},
	}
	jsonData, _ := json.Marshal(payload)
	return bytes.NewBuffer(jsonData)
}

type DiscordClient struct {
	webhookUrl string
	httpClient *http.Client
}

func NewDiscordClient(webhookUrl string) *DiscordClient {
	return &DiscordClient{webhookUrl: webhookUrl, httpClient: &http.Client{}}
}

func (client *DiscordClient) Trigger(task *Task) {
	payload := makePayloadBuffer(task)
	method := "POST"
	url := client.webhookUrl
	if task.WebhookId != nil {
		method = "PATCH"
		url = client.webhookUrl + "/messages/" + *task.WebhookId
	}
	req, _ := http.NewRequest(method, url, payload)
	req.Header.Set("Content-Type", "application/json")
	req.URL.RawQuery = "wait=true"
	resp, _ := client.httpClient.Do(req)
	defer resp.Body.Close()
	data := DiscordResponse{}
	_ = json.NewDecoder(resp.Body).Decode(&data)
	task.WebhookId = &data.Id
}

File: lib/src/discord.rs
use reqwest::{Method, Response};
use serde::{Deserialize, Serialize};
use tokio::{self, sync::broadcast::Receiver};

use crate::ffmpeg::FFMpegEvent;

#[derive(Serialize)]
pub struct DiscordEmbedField {
    name: String,
    value: String,
    inline: Option<bool>,
}

#[derive(Serialize)]
pub struct DiscordEmbed {
    title: Option<String>,
    description: Option<String>,
    color: Option<u32>,
    fields: Option<Vec<DiscordEmbedField>>,
}

#[derive(Serialize)]
pub struct DiscordWebhookData {
    embeds: Vec<DiscordEmbed>,
}

#[derive(Deserialize)]
pub struct DiscordWebhookResponse {
    id: String,
}

#[derive(Clone)]
pub struct DiscordWebhook {
    url: String,
}

impl DiscordWebhook {
    pub fn new(url: &String) -> DiscordWebhook {
        DiscordWebhook { url: url.clone() }
    }

    async fn fetch(&self, method: Method, path: &str, data: DiscordEmbed) -> Response {
        let url = self.url.to_owned() + path;
        reqwest::Client::new()
            .request(method, url)
            .json(&DiscordWebhookData { embeds: vec![data] })
            .send()
            .await
            .expect("Discord fetch failed")
    }

    async fn execute(&self, embed: DiscordEmbed) -> DiscordWebhookMessage {
        let response = self.fetch(Method::POST, "?wait=true", embed).await;

        let data = response
            .json::<DiscordWebhookResponse>()
            .await
            .expect("Invalid json response");

        DiscordWebhookMessage {
            id: data.id,
            webhook: self.clone(),
        }
    }
}

#[derive(Clone)]
pub struct DiscordWebhookMessage {
    id: String,
    webhook: DiscordWebhook,
}

impl DiscordWebhookMessage {
    async fn update(&self, embed: DiscordEmbed) {
        self.webhook
            .fetch(
                Method::PATCH,
                format!("/messages/{}", self.id).as_str(),
                embed,
            )
            .await;
    }
}

pub struct DiscordEventHandler {
    webhook: DiscordWebhook,
}

impl DiscordEventHandler {
    pub fn new(webhook: DiscordWebhook) -> Self {
        Self { webhook }
    }

    fn get_payload(event: &FFMpegEvent) -> Option<DiscordEmbed> {
        let (description, color, context, mut additional_fields) = match event {
            FFMpegEvent::START(context) => {
                ("Waiting for ffmpeg to start...", 0xa855f7, context, vec![])
            }
            FFMpegEvent::PROGRESS(context, progress) => (
                "Transcoding file...",
                0xf97316,
                context,
                vec![
                    DiscordEmbedField {
                        name: "Duration".into(),
                        value: format!(
                            "{:.0}s",
                            context.probe.format.duration.parse::<f64>().unwrap()
                        ),
                        inline: Some(true),
                    },
                    DiscordEmbedField {
                        name: "Timestamp".into(),
                        value: format!("{}s", progress.out_time_us / 1_000_000),
                        inline: Some(false),
                    },
                    DiscordEmbedField {
                        name: "Speed".into(),
                        value: progress.speed.clone(),
                        inline: Some(false),
                    },
                ],
            ),
            FFMpegEvent::DONE(context) => {
                ("Transcoded file successfully", 0x22c55e, context, vec![])
            }
            FFMpegEvent::ERROR(context) => {
                ("An unexpected error happened", 0xef4444, context, vec![])
            }
            _ => return None,
        };

        let mut fields = vec![
            DiscordEmbedField {
                name: "Input".into(),
                value: context.input_path.clone(),
                inline: Some(false),
            },
            DiscordEmbedField {
                name: "Output".into(),
                value: context.output_path.clone(),
                inline: Some(false),
            },
        ];

        fields.append(&mut additional_fields);

        fields.push(DiscordEmbedField {
            name: "Command".into(),
            value: format!("```shell\n{}\n```", context.command),
            inline: Some(false),
        });

        Some(DiscordEmbed {
            title: Some("Transcoding file".into()),
            description: Some(description.into()),
            color: Some(color),
            fields: Some(fields),
        })
    }

    pub async fn listen(&mut self, mut rx: Receiver<FFMpegEvent>) {
        let mut message_opt: Option<DiscordWebhookMessage> = None;

        while let Ok(event) = rx.recv().await {
            if let Some(embed) = Self::get_payload(&event) {
                if let FFMpegEvent::START(_) = event {
                    message_opt = Some(self.webhook.clone().execute(embed).await);
                } else if let Some(message) = &message_opt {
                    message.update(embed).await;
                }
            }

            if let FFMpegEvent::CLOSE() = event {
                break;
            }
        }
    }
}
